File: ImperatorToCK3/Source/Imperator/Provinces/Provinces.cpp
#include "Provinces.h"
#include "Province.h"
#include "Pops.h"
#include "../Countries/Countries.h"
#include "../Countries/Country.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"
#include <ranges>



Imperator::Provinces::Provinces(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::Provinces::registerKeys() {
	registerRegex(commonItems::integerRegex, [this](const std::string& provID, std::istream& theStream) {
		std::shared_ptr<Province> newProvince = provinceFactory.getProvince(theStream, commonItems::stringToInteger<unsigned long long>(provID));
		provinces.emplace(newProvince->getID(), newProvince);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}


void Imperator::Provinces::linkPops(const Pops& thePops) {
	auto counter = 0;
	const auto& pops = thePops.getPops();
	for (const auto& province : provinces | std::views::values) {
		if (!province->getPops().empty()) {
			std::map<unsigned long long, std::shared_ptr<Pop>> newPops;
			for (const auto& popID : province->getPops() | std::views::keys) {
				const auto& popItr = pops.find(popID);
				if (popItr != pops.end()) {
					newPops.emplace(popItr->first, popItr->second);
					++counter;
				}
				else {
					Log(LogLevel::Warning) << "Pop ID: " << popID << " has no definition!";
				}
			}
			province->setPops(newPops);
		}
	}
	Log(LogLevel::Info) << "<> " << counter << " pops linked to provinces.";
}


void Imperator::Provinces::linkCountries(const Countries& theCountries) {
	auto counter = 0;
	const auto& countries = theCountries.getCountries();
	for (const auto& province : provinces | std::views::values) {
		if (!province->getPops().empty()) {
			const auto& countryItr = countries.find(province->getOwner().first);
			if (countryItr != countries.end()) {
				// link both ways
				province->linkOwnerCountry(countryItr->second);
				countryItr->second->registerProvince(province);
				++counter;
			}
			else {
				Log(LogLevel::Warning) << "Country ID: " << province->getOwner().first << " has no definition!";
			}
		}
	}
	Log(LogLevel::Info) << "<> " << counter << " countries linked to provinces.";
}

File: ImperatorToCK3/Imperator/Provinces/Provinces.cs
using commonItems;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;

namespace ImperatorToCK3.Imperator.Provinces {
	public class Provinces : IReadOnlyDictionary<ulong, Province> {
		public Provinces() { }
		public Provinces(BufferedReader reader) {
			var parser = new Parser();
			RegisterKeys(parser);
			parser.ParseStream(reader);
		}
		private void RegisterKeys(Parser parser) {
			parser.RegisterRegex(CommonRegexes.Integer, (reader, provIdStr) => {
				var newProvince = Province.Parse(reader, ulong.Parse(provIdStr));
				Add(newProvince.Id, newProvince);
			});
			parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreAndLogItem);
		}
		public void LinkPops(Pops.Pops pops) {
			var counter = Values.Sum(province => province.LinkPops(pops));
			Logger.Info($"{counter} pops linked to provinces.");
		}
		public void LinkCountries(Countries.Countries countries) {
			var counter = 0;
			foreach (var province in Values) {
				if (countries.TryGetValue(province.OwnerCountry.Key, out var countryToLink)) {
					// link both ways
					province.LinkOwnerCountry(countryToLink);
					countryToLink.RegisterProvince(province);
					++counter;
				} else {
					Logger.Warn($"Country with ID {province.OwnerCountry.Key} has no definition!");
				}
			}
			Logger.Info($"{counter} countries linked to provinces.");
		}

		public bool ContainsKey(ulong key) => provincesDict.ContainsKey(key);
		public bool TryGetValue(ulong key, [MaybeNullWhen(false)] out Province value) => provincesDict.TryGetValue(key, out value);
		public IEnumerator<KeyValuePair<ulong, Province>> GetEnumerator() => provincesDict.GetEnumerator();
		IEnumerator IEnumerable.GetEnumerator() => provincesDict.GetEnumerator();
		public IEnumerable<ulong> Keys => provincesDict.Keys;
		public IEnumerable<Province> Values => provincesDict.Values;
		public int Count => provincesDict.Count;
		public Province this[ulong key] => provincesDict[key];
		private readonly Dictionary<ulong, Province> provincesDict = new();
	}
}
